### ISSUE TITLE
The case sensitive issue when filtering data with string.Contains method in both in memory and SQL Server.

### ISSUE THREAD

**Issue by RickyLin** [dotnet/efcore#1353]
I need EF 7 to generate statements similar to `LIKE '%' + str + '%'`, so I implemented a ContainsTranslator. I also created tests for it. However, the test results were different  when running in memory test and SQL Server test because the string.Contains method is case sensitive in in memory (19 entries) and its translation `LIKE '%' + str + '%'` is case insensitive in SQL Server (34 entries). How should we address this to pass the test?

**Comment by rowanmiller** [comment]
@RickyLin - For other operations like this we are currently applying the in-memory filter in addition to the database filter. This means folks get results that are consistent with the CLR behavior of the method/comparison (which is typically what people want). Of course, this means the database may return more rows than are ultimately returned from the LINQ query. If we find that performance is unacceptable in these situations we may have a flag to disable it.

**Comment by rowanmiller** [comment]
FYI the issue tracking this is https://github.com/aspnet/EntityFramework/issues/474
Let us know if you are thinking about contributing a pull request for this :smile:

**Comment by RickyLin** [comment]
@rowanmiller - I think my problem is related to #474. But I'm not trying to solve the escaping issue in my case. :) 
By the way I'm not an English native speaker :-( So sorry for misunderstanding if any.

I didn't see an in-memory filter is applied in addition to the database filter. I modified the QueryTestBase.cs, simply replaced "M" with "m" and 12 with 0, then ran the test.

``` c#
    [Fact]
    public virtual void String_StartsWith_Literal()
    {
        AssertQuery<Customer>(
            cs => cs.Where(c => c.ContactName.StartsWith("m")),
            stateEntryCount: 0);
    }
```

the EntityFramework.InMemory.FunctionalTests succeeded since all contact names start with "M" but EntityFramework.SqlServer.FunctionalTests failed because the Sql Server returned 12 entries which was correct for Sql Server, it didn't seem an in-memory filter was working on the result Sql Server returned. This can be worked around by that no contact name starts with "m" in test data for now. But I can't take advantage of this workaround in my Contains test method. My point is that whether I can work around the same situation in my Contains test method by this:
the test method in QueryTestBase.cs

```c#
    [Fact]
    public virtual void String_Contains_Literal()
    {
        AssertQuery<Customer>(
            cs => cs.Where(c => c.ContactName.Contains("M")),
            stateEntryCount: 19);
    }
```
so EntityFramework.InMemory.FunctionalTests can succeed.

and in SqlServerQueryTest.cs

```c#
    public override void String_Contains_Literal()
    {
        // base.String_Contains_Literal(); // commented out, so it won't check if the entity count is 0

        Assert.Equal(
            @"SELECT [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[CustomerID], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE [c].[ContactName] LIKE '%' + @p0 + '%'",
            Sql);
    }
```
so EntityFramework.SqlServer.FunctionalTests don't care about the entity count from Sql Server is 34 and just care about the generated sql statement.

If I can do this, I think I can create a pull request so we can make the Contains method of string work in Where clause.

**Comment by rowanmiller** [comment]
Makes sense, at one point we were actually still applying the in-memory filter after bringing results back from the database for operations such as `.Where(c => c.Name == "Fred")`, but it looks like that is no longer the case (sorry for the randomization).

@maumar should be able to help you out with how to handle this appropriately in the tests. I'm assuming you probably just want to override the test method in the test class and change the expected result count.

**Comment by maumar** [comment]
One way of handling this could be adding overloads to AssertQuery that would take two funcs, e.g.

```
        protected void AssertQuery<TItem>(
            Func<IQueryable<TItem>, IQueryable<object>> productQuery,
            Func<IQueryable<TItem>, IQueryable<object>> oracleQuery,
            bool assertOrder = false,
            int stateEntryCount = 0)
            where TItem : class
        {
            using (var context = CreateContext())
            {
                AssertResults(
                    productQuery(NorthwindData.Set<TItem>()).ToArray(),
                    oracleQuery(context.Set<TItem>()).ToArray(),
                    assertOrder);

                Assert.Equal(stateEntryCount, context.ChangeTracker.Entries().Count());
            }
        }
```

this way you can provide different query that would produce expected results and compensate for the difference between CLR and sql server (e.g add ToLower() calls in the comparison)

**Comment by maumar** [comment]
Although this would also require abstracting out the oracleQuery for this particular test between providers, so that for InMemory it does basic filter (case sensitive) and for SqlServer it does case insensitive filter (using the ToLower() etc).

**Comment by rowanmiller** [comment]
I see we have a pull request for this now, so we can continue any discussion about test patterns on the review of the proposed code.

**Comment by rowanmiller** [comment]
BTW - @RickyLin thanks for reaching out to us about this one... it's always great to have proactive discussions about things like this before we get a pull request... appreciate it!

**Comment by maumar** [comment]
merged in 76ad8c149d17ce698333e502d78509be1d4327e6 and b92eb1efa5cb46116197a8dde3d245afb243cab5
